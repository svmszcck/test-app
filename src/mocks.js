import React from "react";
import configureMockStore from "redux-mock-store";
import * as redux from "react-redux";
import { NavigationContainer } from "@react-navigation/native";
import { createStackNavigator } from "@react-navigation/stack";
import thunk from "redux-thunk";

const mockStore = configureMockStore([thunk]);
const Stack = createStackNavigator();

export const store = mockStore({
  user: {
    name: "Onur",
    avatar: null,
    favorites: [],
  },
  posts: {
    movie: {
      id: 12,
      title: "Rwer",
      poster_path: "ekjrhtkjert",
      vote_average: 5.6,
    },
  },
});

export const MockedNavigator = ({ component, params = {} }) => {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen
          name="MockedScreen"
          component={component}
          initialParams={params}
        />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export const navigation = {
  navigate: jest.fn(),
};

export const mockUseSelector = jest.fn();
export const mockUseDispatch = jest.fn();

export const useDispatchSpy = jest.spyOn(redux, "useDispatch");
export const mockDispatchFn = jest.fn();
useDispatchSpy.mockReturnValue(mockDispatchFn);
