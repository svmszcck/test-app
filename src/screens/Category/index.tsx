import React, { useEffect, useMemo, useState } from "react";
import { useDispatch, useSelector } from "react-redux";

import { getMovieByGenre, resetMoviesByGenre } from "store/actions/posts";
import { Store } from "types";
import CategoryView from "./view";

const Category = ({ navigation, route }: CategoryProps) => {
  const dispatch = useDispatch();
  const [page, setPage] = useState<number>(1);
  const posts = useSelector((state: Store) => state.posts);
  const { moviesByGenre, moviesByGenreLoading } = posts;

  const id = useMemo(() => route?.params?.id, [route]);

  const name = useMemo(() => route?.params?.name, [route]);

  useEffect(() => {
    if (id) dispatch(getMovieByGenre(id, page));
    return () => {
      dispatch(resetMoviesByGenre());
    };
  }, []);

  const loadMore = () => {
    if (id) {
      dispatch(getMovieByGenre(id, page + 1));
      setPage(page + 1);
    }
  };

  return (
    <CategoryView
      navigation={navigation}
      movies={moviesByGenre}
      loading={moviesByGenreLoading}
      loadMore={loadMore}
      name={name}
    />
  );
};

type CategoryProps = {
  navigation: any;
  route: {
    params: any;
  };
};

export default Category;
