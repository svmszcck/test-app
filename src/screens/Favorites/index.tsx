import React from "react";
import { Alert } from "react-native";
import { useDispatch, useSelector } from "react-redux";

import { toggleFavorite } from "store/actions/user";
import { Store } from "types";
import FavoritesView from "./view";

const Favorites = ({ navigation }: FavoritesProps) => {
  const dispatch = useDispatch();
  const user = useSelector((state: Store) => state.user);
  const { favorites } = user;

  const removeItem = (id: number) => {
    Alert.alert(
      "Warning!",
      "Are you sure you want to remove this movie from your favorite list",
      [
        {
          text: "Cancel",
          onPress: () => null,
          style: "cancel",
        },
        { text: "Yes", onPress: () => dispatch(toggleFavorite({ id })) },
      ]
    );
  };

  return (
    <FavoritesView
      favorites={favorites}
      removeItem={removeItem}
      navigation={navigation}
    />
  );
};

type FavoritesProps = {
  navigation: any;
};

export default Favorites;
