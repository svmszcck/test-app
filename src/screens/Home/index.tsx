import React, { useEffect, useState } from "react";
import { Alert, BackHandler, Keyboard } from "react-native";
import { useDispatch, useSelector } from "react-redux";

import useBackHandler from "hooks/useBackHandler";
import { getGenres, getPopularMovies, resetPosts } from "store/actions/posts";
import { handleBackButton, showSurprise } from "utils/ui";
import { Store } from "types";
import HomeView from "./view";

const Home = ({ navigation }: HomeProps) => {
  const [categories, setCategories] = useState<boolean>(false);
  const [page, setPage] = useState<number>(1);
  const [isRefreshing, showRefresh] = useState<boolean>(false);
  const posts = useSelector((state: Store) => state.posts);
  const user = useSelector((state: Store) => state.user);
  const dispatch = useDispatch();
  const {
    genres,
    popularMovies,
    genresLoading,
    moviesLoading,
    popularMoviesLoading,
  } = posts;
  const { name } = user;

  useEffect(() => {
    dispatch(getGenres());
    dispatch(getPopularMovies(page));
  }, []);

  const loadPopularMovies = () => {
    dispatch(getPopularMovies(page + 1));
    setPage(page + 1);
  };

  const refreshPage = () => {
    showRefresh(true);
    dispatch(resetPosts());
    dispatch(getGenres());
    dispatch(getPopularMovies(1));
    setPage(1);
    showRefresh(false);
  };

  useBackHandler(() => {
    handleBackButton();
    return true;
  });

  return (
    <HomeView
      genres={genres}
      popularMovies={popularMovies}
      navigation={navigation}
      categories={categories}
      setCategories={setCategories}
      loadMore={loadPopularMovies}
      genresLoading={genresLoading}
      moviesLoading={moviesLoading}
      popularMoviesLoading={popularMoviesLoading}
      isRefreshing={isRefreshing}
      refresh={refreshPage}
      name={name}
      showMessage={() => showSurprise(name)}
    />
  );
};

type HomeProps = {
  navigation: any;
};

export default Home;
