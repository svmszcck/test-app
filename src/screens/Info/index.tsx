import React from "react";
import { View } from "react-native";

const Info = () => {
  return <View></View>;
};

export default Info;
