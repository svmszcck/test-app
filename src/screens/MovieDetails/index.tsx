import React, { useEffect, useMemo } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useFocusEffect } from "@react-navigation/native";

import { getMovieDetails, resetMovie } from "store/actions/posts";
import { toggleFavorite } from "store/actions/user";
import { openLink } from "utils/general";
import { IMDB_URL } from "app_constants/api";
import { Movie, Store } from "types";
import MovieDetailsView from "./view";

const MovieDetails = ({ navigation, route }: MovieDetailsProps) => {
  const dispatch = useDispatch();
  const posts = useSelector((state: Store) => state.posts);
  const user = useSelector((state: Store) => state.user);
  const { movie, movieLoading } = posts;
  const {
    id,
    title,
    poster_path: posterPath,
    vote_average: voteAverage,
  }: Movie = movie;
  const { favorites } = user;

  const isFavorite = useMemo(
    () => favorites.find((favorite: any) => favorite?.id === id),
    [favorites, id]
  );

  useEffect(() => {
    const { id } = route.params;
    if (movie.id !== id) {
      dispatch(resetMovie());
      dispatch(getMovieDetails(id));
    }
  }, []);

  const doToggleFavorite = () => {
    dispatch(
      toggleFavorite({
        id,
        title,
        poster_path: posterPath,
        vote_average: voteAverage,
      })
    );
  };

  const openIMDB = (id: string) => openLink(`${IMDB_URL}/${id}`);

  return (
    <MovieDetailsView
      navigation={navigation}
      toggleFavorite={doToggleFavorite}
      movie={movie}
      movieLoading={movieLoading}
      isFavorite={isFavorite}
      openIMDB={openIMDB}
    />
  );
};

type MovieDetailsProps = {
  navigation: any;
  route: {
    params: any;
  };
};

export default MovieDetails;
