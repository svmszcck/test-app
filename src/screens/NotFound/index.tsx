import { StackScreenProps } from "@react-navigation/stack";
import * as React from "react";
import { StyleSheet, Text, TouchableOpacity, View } from "react-native";

import Routes from "app_constants/routes";
import { RootStackParamList } from "types";
import { useColor } from "hooks";

const NotFound = ({
  navigation,
}: StackScreenProps<RootStackParamList, Routes.NOT_FOUND>) => {
  const colors = useColor();
  return (
    <View style={[styles.container, { backgroundColor: colors.white }]}>
      <Text style={styles.title}>Something went wrong :(</Text>
      <TouchableOpacity
        onPress={() => navigation.replace(Routes.ROOT)}
        style={styles.link}
      >
        <Text style={[styles.linkText, { color: colors.link }]}>
          Go to home screen!
        </Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
    padding: 20,
  },
  title: {
    fontSize: 20,
    fontWeight: "bold",
  },
  link: {
    marginTop: 15,
    paddingVertical: 15,
  },
  linkText: {
    fontSize: 14,
  },
});

export default NotFound;
