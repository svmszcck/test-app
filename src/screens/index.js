export { default as Home } from "./Home";
export { default as Category } from "./Category";
export { default as MovieDetails } from "./MovieDetails";
export { default as Welcome } from "./Welcome";
export { default as Profile } from "./Profile";
export { default as Favorites } from "./Favorites";
export { default as Search } from "./Search";
export { default as NotFound } from "./NotFound";
