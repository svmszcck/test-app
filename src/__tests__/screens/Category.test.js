import "react-native";
import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";

import { Category } from "screens";
import { store } from "mocks";

jest.useFakeTimers();

it("renders correctly", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Category />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
