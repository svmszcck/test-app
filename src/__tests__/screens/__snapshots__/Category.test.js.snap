// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Category Screen renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "padding": 20,
        "paddingBottom": 0,
      },
      Object {
        "backgroundColor": "#E4F5F3",
      },
    ]
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 20,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "opacity": 1,
        }
      }
    >
      <Text />
      <Text
        style={
          Object {
            "color": "#5C7B8A",
            "fontSize": 18,
            "marginLeft": 10,
          }
        }
      >
        Back
      </Text>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginLeft": "auto",
          "opacity": 1,
          "width": 30,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "marginBottom": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#0B354F",
          "fontSize": 25,
          "marginBottom": 10,
        }
      }
    >
      
    </Text>
    <View>
      <RCTScrollView
        contentContainerStyle={
          Object {
            "paddingBottom": 40,
          }
        }
        showsVerticalScrollIndicator={false}
      >
        <View>
          <RCTScrollView
            data={
              Array [
                Object {
                  "id": 12,
                  "poster_path": "rtyrty",
                  "title": "Hehehe",
                  "vote_average": 7.6,
                },
              ]
            }
            disableVirtualization={false}
            getItem={[Function]}
            getItemCount={[Function]}
            horizontal={false}
            initialNumToRender={10}
            keyExtractor={[Function]}
            maxToRenderPerBatch={10}
            onContentSizeChange={[Function]}
            onEndReachedThreshold={2}
            onLayout={[Function]}
            onMomentumScrollEnd={[Function]}
            onScroll={[Function]}
            onScrollBeginDrag={[Function]}
            onScrollEndDrag={[Function]}
            removeClippedSubviews={false}
            renderItem={[Function]}
            scrollEventThrottle={50}
            stickyHeaderIndices={Array []}
            updateCellsBatchingPeriod={50}
            viewabilityConfigCallbackPairs={Array []}
            windowSize={21}
          >
            <View>
              <View
                onLayout={[Function]}
                style={null}
              >
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "marginBottom": 12,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "#ffffff",
                        "borderColor": "#bcbbc1",
                        "flexDirection": "row",
                        "padding": 14,
                      }
                    }
                    testID="padView"
                  >
                    <View
                      style={
                        Object {
                          "backgroundColor": "transparent",
                          "height": 75,
                          "width": 75,
                        }
                      }
                      theme={
                        Object {
                          "colors": Object {
                            "black": "#242424",
                            "disabled": "hsl(208, 8%, 90%)",
                            "divider": "#bcbbc1",
                            "error": "#ff190c",
                            "grey0": "#393e42",
                            "grey1": "#43484d",
                            "grey2": "#5e6977",
                            "grey3": "#86939e",
                            "grey4": "#bdc6cf",
                            "grey5": "#e1e8ee",
                            "greyOutline": "#bbb",
                            "platform": Object {
                              "android": Object {
                                "error": "#f44336",
                                "grey": "rgba(0, 0, 0, 0.54)",
                                "primary": "#2196f3",
                                "secondary": "#9C27B0",
                                "success": "#4caf50",
                                "warning": "#ffeb3b",
                              },
                              "ios": Object {
                                "error": "#ff3b30",
                                "grey": "#7d7d7d",
                                "primary": "#007aff",
                                "searchBg": "#dcdce1",
                                "secondary": "#5856d6",
                                "success": "#4cd964",
                                "warning": "#ffcc00",
                              },
                            },
                            "primary": "#2089dc",
                            "searchBg": "#303337",
                            "secondary": "#ca71eb",
                            "success": "#52c41a",
                            "warning": "#faad14",
                            "white": "#ffffff",
                          },
                        }
                      }
                    >
                      <View
                        accessibilityIgnoresInvertColors={true}
                        style={
                          Object {
                            "backgroundColor": "transparent",
                            "flex": 1,
                            "overflow": "hidden",
                            "position": "relative",
                          }
                        }
                      >
                        <Image
                          onLoad={[Function]}
                          source={
                            Object {
                              "uri": "http://image.tmdb.org/t/p/w200/rtyrty",
                            }
                          }
                          style={
                            Object {
                              "bottom": 0,
                              "flex": 1,
                              "height": null,
                              "left": 0,
                              "position": "absolute",
                              "right": 0,
                              "top": 0,
                              "width": null,
                            }
                          }
                          testID="RNE__Image"
                          theme={
                            Object {
                              "colors": Object {
                                "black": "#242424",
                                "disabled": "hsl(208, 8%, 90%)",
                                "divider": "#bcbbc1",
                                "error": "#ff190c",
                                "grey0": "#393e42",
                                "grey1": "#43484d",
                                "grey2": "#5e6977",
                                "grey3": "#86939e",
                                "grey4": "#bdc6cf",
                                "grey5": "#e1e8ee",
                                "greyOutline": "#bbb",
                                "platform": Object {
                                  "android": Object {
                                    "error": "#f44336",
                                    "grey": "rgba(0, 0, 0, 0.54)",
                                    "primary": "#2196f3",
                                    "secondary": "#9C27B0",
                                    "success": "#4caf50",
                                    "warning": "#ffeb3b",
                                  },
                                  "ios": Object {
                                    "error": "#ff3b30",
                                    "grey": "#7d7d7d",
                                    "primary": "#007aff",
                                    "searchBg": "#dcdce1",
                                    "secondary": "#5856d6",
                                    "success": "#4cd964",
                                    "warning": "#ffcc00",
                                  },
                                },
                                "primary": "#2089dc",
                                "searchBg": "#303337",
                                "secondary": "#ca71eb",
                                "success": "#52c41a",
                                "warning": "#faad14",
                                "white": "#ffffff",
                              },
                            }
                          }
                          transition={true}
                        />
                        <View
                          accessibilityElementsHidden={true}
                          importantForAccessibility="no-hide-descendants"
                          pointerEvents="none"
                          style={
                            Object {
                              "bottom": 0,
                              "left": 0,
                              "opacity": 1,
                              "position": "absolute",
                              "right": 0,
                              "top": 0,
                            }
                          }
                        >
                          <View
                            style={
                              Object {
                                "alignItems": "center",
                                "backgroundColor": "#bdbdbd",
                                "flex": 1,
                                "height": null,
                                "justifyContent": "center",
                                "width": null,
                              }
                            }
                            testID="RNE__Image__placeholder"
                          />
                        </View>
                        <View
                          style={
                            Object {
                              "flex": 1,
                              "height": null,
                              "width": null,
                            }
                          }
                        />
                      </View>
                    </View>
                    <View
                      style={
                        Object {
                          "paddingLeft": 16,
                        }
                      }
                    />
                    <View
                      style={
                        Object {
                          "alignItems": "flex-start",
                          "flex": 1,
                          "justifyContent": "center",
                        }
                      }
                      theme={
                        Object {
                          "colors": Object {
                            "black": "#242424",
                            "disabled": "hsl(208, 8%, 90%)",
                            "divider": "#bcbbc1",
                            "error": "#ff190c",
                            "grey0": "#393e42",
                            "grey1": "#43484d",
                            "grey2": "#5e6977",
                            "grey3": "#86939e",
                            "grey4": "#bdc6cf",
                            "grey5": "#e1e8ee",
                            "greyOutline": "#bbb",
                            "platform": Object {
                              "android": Object {
                                "error": "#f44336",
                                "grey": "rgba(0, 0, 0, 0.54)",
                                "primary": "#2196f3",
                                "secondary": "#9C27B0",
                                "success": "#4caf50",
                                "warning": "#ffeb3b",
                              },
                              "ios": Object {
                                "error": "#ff3b30",
                                "grey": "#7d7d7d",
                                "primary": "#007aff",
                                "searchBg": "#dcdce1",
                                "secondary": "#5856d6",
                                "success": "#4cd964",
                                "warning": "#ffcc00",
                              },
                            },
                            "primary": "#2089dc",
                            "searchBg": "#303337",
                            "secondary": "#ca71eb",
                            "success": "#52c41a",
                            "warning": "#faad14",
                            "white": "#ffffff",
                          },
                        }
                      }
                    >
                      <Text
                        style={
                          Object {
                            "backgroundColor": "transparent",
                            "color": "#242424",
                            "fontSize": 17,
                          }
                        }
                        testID="listItemTitle"
                      >
                        Hehehe
                      </Text>
                      <Text
                        style={
                          Object {
                            "backgroundColor": "transparent",
                            "color": "#5C7B8A",
                            "fontSize": 15,
                          }
                        }
                        testID="listItemTitle"
                      >
                        Score: 7.6
                      </Text>
                    </View>
                    <View
                      style={
                        Object {
                          "paddingLeft": 16,
                        }
                      }
                    />
                    <View
                      style={
                        Object {
                          "alignSelf": "center",
                          "overflow": "hidden",
                        }
                      }
                    >
                      <View>
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "backgroundColor": "transparent",
                              "justifyContent": "center",
                            }
                          }
                        >
                          <Text />
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </RCTScrollView>
          <View
            style={
              Object {
                "alignItems": "center",
                "height": 80,
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "borderRadius": 3,
                  "overflow": "hidden",
                }
              }
            >
              <View
                accessibilityRole="button"
                accessibilityState={
                  Object {
                    "busy": false,
                    "disabled": false,
                  }
                }
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#2CA17C",
                      "borderColor": "#2089dc",
                      "borderRadius": 3,
                      "borderWidth": 0,
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "margin": 20,
                      "padding": 8,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "white",
                        "fontSize": 18,
                        "paddingBottom": 1,
                        "paddingTop": 2,
                        "textAlign": "center",
                      }
                    }
                  >
                    Load More
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;
